File: rules/queries/printAST.pl
:- use_module(library(http/json)).
:- use_module(library(http/json_convert)).


is_declaration(Node, KindAtom, Name, IdNode) :-
    declaration_node(Node, KindAtom, IdNode),
    identifier_name(IdNode, Name).

% is_reference(Node, Name, IdNode):
%   Checks if Node is recognized by the adapter as a reference, then
%   retrieves the identifier Name.
is_reference(Node, Name, IdNode) :-
    reference_node(Node, IdNode),
    identifier_name(IdNode, Name).

% --------------------------------------------------------------------------
% build_ast/2
%   Recursively constructs a JSON-like representation of Node:
%     - nodeName: always the raw ANTLR type (e.g., "ProgramContext")
%     - nodeKind: "declaration", "reference", or null
%     - symbolKind: e.g., "function" or "variable" (only for declarations)
%     - name: the identifier (only for references and declarations)
%     - location: the node's File/Line/Col/Len (only for references/declarations)
%     - children: recursively built AST children
% --------------------------------------------------------------------------

build_ast(Node, Ast) :-
    % 1) If Node is a declaration:
    is_declaration(Node, SymbolKind, Name, IdNode),
    node(Node, AntlrType),
    node_location(IdNode, File, Line, Col, Len),
    findall(ChildAst, (child(Node, Child), build_ast(Child, ChildAst)), ChildrenAsts),
    Ast = json{
        nodeName: AntlrType,
        nodeKind: "declaration",
        symbolKind: SymbolKind,   % e.g. "function", "variable"
        name: Name,
        location: json{file: File, line: Line, col: Col, len: Len},
        children: ChildrenAsts
    },
    !.  % Cut to prevent falling into later clauses.

build_ast(Node, Ast) :-
    % 2) If Node is a reference:
    is_reference(Node, Name, IdNode),
    node(Node, AntlrType),
    node_location(IdNode, File, Line, Col, Len),
    findall(ChildAst, (child(Node, Child), build_ast(Child, ChildAst)), ChildrenAsts),
    Ast = json{
        nodeName: AntlrType,
        nodeKind: "reference",
        name: Name,
        location: json{file: File, line: Line, col: Col, len: Len},
        children: ChildrenAsts
    },
    !.

build_ast(Node, Ast) :-
    % 3) Fallback: no known declaration or reference → nodeKind: null
    node(Node, AntlrType),
    findall(ChildAst, (child(Node, Child), build_ast(Child, ChildAst)), ChildrenAsts),
    Ast = json{
        nodeName: AntlrType,
        nodeKind: null,
        children: ChildrenAsts
    }.

% --------------------------------------------------------------------------
% main_ast/0
%   Finds top-level root nodes (often "ProgramContext"), builds the AST for
%   each, prints them as JSON (list form if multiple roots), then halts.
% --------------------------------------------------------------------------

main_ast :-
    % Typically there's one "ProgramContext", but in case of multiple roots:
    findall(Root, node(Root, 'ProgramContext'), Roots),
    maplist(build_ast, Roots, RootAsts),
    json_write_dict(current_output, RootAsts, [width(80)]),
    halt.

:- main_ast.

File: rules/queries/schemas/printAST.ts
import { z } from 'zod';

// 1) Location schema (similar to LSP's `Location`)
const LocationSchema = z.object({
  file: z.string(),
  line: z.number(),
  col: z.number(),
  len: z.number(),
});

// 2) Enum-like schema for `nodeKind`
const NodeKindSchema = z.union([
  z.literal("declaration"),
  z.literal("reference"),
  z.null(),
]);

// 3) Define TypeScript type for AST node
type ASTNode = {
  nodeName: string;
  nodeKind: "declaration" | "reference" | null;
  symbolKind?: string; // e.g., "function", "variable", "class"...
  name?: string;
  location?: z.infer<typeof LocationSchema>;
  children: ASTNode[];
};

// 4) Define Zod schema using a manual type hint
const ASTNodeSchema: z.ZodType<ASTNode> = z.object({
  nodeName: z.string(),
  nodeKind: NodeKindSchema,
  symbolKind: z.string().optional(),
  name: z.string().optional(),
  location: LocationSchema.optional(),
  children: z.lazy(() => ASTNodeSchema.array()), // Recursive reference
});

// 5) The full AST is an array of root-level nodes
const ASTSchema = z.array(ASTNodeSchema);

export { ASTNodeSchema, ASTSchema, LocationSchema };
export type { ASTNode };